### PR TITLE
Add layout option to woq int4 api

### PR DESCRIPTION
**Pull request by Diogo-V**
### Summary
This PR updates the `int4_weight_only` API by introducing a `layout_type` parameter. This will allow users to quantize models to int4 while being able to choose between available layouts.

This change also breaks backwards compatibility for all users that explicitly define the `inner_k_tiles` parameter:
```py
# old
def int4_weight_only(group_size=128, inner_k_tiles=8):
...

quantize_(my_model, int4_weight_only(inner_k_tiles=8))

# new
def int4_weight_only(group_size=128, layout_type=TensorCoreTiledLayoutType(inner_k_tiles=8)):
...

quantize_(my_model, int4_weight_only(layout_type=TensorCoreTiledLayoutType(inner_k_tiles=8)))
```

### BC Breaking notes for Release statement

* `inner_k_tiles` was deprecated in favor of `layout_type`, enabling users to select from various layout options [#670](https://github.com/pytorch/ao/pull/670)

```py
# for torchao 0.5
from torchao.quantization import quantize, int4_weight_only
quantize_(my_model, int4_weight_only(layout_type=TensorCoreTiledLayoutType(inner_k_tiles=8)))

# for torchao 0.4
from torchao.quantization import quantize_, int4_weight_only
quantize_(my_model, int4_weight_only(inner_k_tiles=8))
```

### Tasks
 - [x] Update the function's interface
 - [x] Update the documentation
 - [x] Update the tests

### PRs pending on this one
* #621 

### Questions for Reviewers
* I am not familiar with how releases and backwards compatibility is maintained in pytorch and since this is a breaking change, how should we handle delivering it? Is there a specific flag/if statement that I should add somewhere?

Let me know if there is anything that I need to update. I would be happy to perform any necessary changes.

### CODE REVIEW

File: torchao/quantization/quant_api.py [hunk 1]
```diff
@@ -21,7 +21,11 @@
 import torch.nn.functional as F
 from typing import Any, Callable, Union, Dict, Optional
 
-from torchao.dtypes import PlainLayoutType
+from torchao.dtypes import (
+    to_affine_quantized,
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:25]: thanks, if there is no circular dep you can remove the import from other functions as well, e.g. `int8_weight_only`

**Comment by jerryzh168** [comment]
thanks, although this is bc-breaking, but I don't think there are many callsites changing `inner_k_tiles` so we should be fine I think

**Comment by jerryzh168** [comment]
could you write a bc-breaking notes for this? see format in https://github.com/pytorch/ao/releases, something like:

```
# for torchao 0.5
...

# for torchao 0.4
...
```

File: torchao/quantization/quant_api.py (hunk 1 above)
Comment (Diogo-V) [reply torchao/quantization/quant_api.py:26]: done

**Comment by Diogo-V** [comment]
on it! where should I place the bc-breaking notes? as a comment on this PR or someplace in the code?

**Comment by Diogo-V** [comment]
* `inner_k_tiles` was deprecated in favor of `layout_type`, enabling users to select from various layout options [#670](https://github.com/pytorch/ao/pull/670)

```py
# for torchao 0.5
from torchao.quantization import quantize_, int4_weight_only
quantize_(my_model, int4_weight_only(inner_k_tiles=8))

# for torchao 0.4
from torchao.quantization import quantize, int4_weight_only
quantize_(my_model, int4_weight_only(layout_type=TensorCoreTiledLayoutType(inner_k_tiles=8)))
```

**Comment by jerryzh168** [comment]
> on it! where should I place the bc-breaking notes? as a comment on this PR or someplace else?

you can put this in PR summary

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks!

**Comment by Diogo-V** [comment]
done ✅ 
feel free to merge it or let me know if there is anything else that I might need to change

**Comment by msaroufim** [comment]
@jerryzh168 do we need to communicate this change to anyone?

**Comment by jerryzh168** [comment]
> @jerryzh168 do we need to communicate this change to anyone?

seems fine I think, I didn't see people using `inner_k_tiles` yet, but we can probably update huggingface when 0.5 is released

**Pull request merged**